src/lib/provable/gadgets/twisted-curve.ts

```diff
@@ -0,0 +1,451 @@
+  let y3Num = ForeignField.sub(y1y2, ax1x2, f);
+
+  let x1x2y1y2 = ForeignField.mul(x1x2, y1y2, f);
+  let dx1x2y1y2 = ForeignField.mul(Field3.from(d), x1x2y1y2, f);
+  // check denominators are correctly computed:
+  // den = 1 / (1 +- d * x1^2 * y1^2)
+  Provable.equal(
+    Field3,
+    one,
+    ForeignField.mul(x3Den, ForeignField.add(one, dx1x2y1y2, f), f)
+  ).assertTrue();
+  Provable.equal(
+    Field3,
+    one,
+    ForeignField.mul(y3Den, ForeignField.sub(one, dx1x2y1y2, f), f)
+  ).assertTrue();
+
+  ForeignField.assertMul(x3Num, x3Den, x3, f);
+  ForeignField.assertMul(y3Num, y3Den, y3, f);
+
+  return { x: x3, y: y3 };
+}
+
+function double(
+  p1: Point,
+  Curve: { modulus: bigint; a: bigint; d: bigint }
+): Point {
+  let { x: x1, y: y1 } = p1;
+  let f = Curve.modulus;
+  let d = Curve.d;
+
+  // constant case
+  if (Point.isConstant(p1)) {
+    let p3 = twistedDouble(Point.toBigint(p1), f, Curve.a, Curve.d);
+    return Point.from(p3);
+  }
+
+  // witness and range-check denominators, x3, y3
+  let witnesses = exists(12, () => {
+    let [x1_, y1_] = Field3.toBigints(x1, y1);
+
+    let x1x1 = mod(x1_ * x1_, f);
+    let y1y1 = mod(y1_ * y1_, f);
+    let x1y1 = mod(x1_ * y1_, f);
+
+    let x3Den = inverse(mod(1n + d * x1x1 * y1y1, f), f) ?? 0n;
+    let y3Den = inverse(mod(1n - d * x1x1 * y1y1, f), f) ?? 0n;
+
+    let x3Num = mod(2n * x1y1, f);
+    let y3Num = mod(y1y1 - Curve.a * x1x1, f);
+
+    let x3 = mod(x3Num * x3Den, f);
+    let y3 = mod(y3Num * y3Den, f);
+
+    return [...split(x3Den), ...split(y3Den), ...split(x3), ...split(y3)];
+  });
+  let [dx0, dx1, dx2, dy0, dy1, dy2, x30, x31, x32, y30, y31, y32] = witnesses;
+  // Note denominators are already inversed
+  let x3Den: Field3 = [dx0, dx1, dx2];
+  let y3Den: Field3 = [dy0, dy1, dy2];
+  let x3: Field3 = [x30, x31, x32];
+  let y3: Field3 = [y30, y31, y32];
+  ForeignField.assertAlmostReduced([x3Den, x3, y3], f);
+  ForeignField.assertAlmostReduced([y3Den], f);
+
+  // x3 = 2*x1*y1 / (1 + d * x1^2 * y1^2)
+  // y3 = (y1^2 - a * x1^2) / (1 - d * x1^2 * y1^2)
+
+  let one = Field3.from(1n);
+  let a = Field3.from(Curve.a);
+  let x1x1 = ForeignField.mul(x1, x1, f);
+  let y1y1 = ForeignField.mul(y1, y1, f);
+  let x1y1 = ForeignField.mul(x1, y1, f);
+  let ax1x1 = ForeignField.mul(a, x1x1, f);
+  let x3Num = ForeignField.add(x1y1, x1y1, f);
+  let y3Num = ForeignField.sub(y1y1, ax1x1, f);
+
+  let x1x1y1y1 = ForeignField.mul(x1x1, y1y1, f);
+  let dx1x1y1y1 = ForeignField.mul(Field3.from(d), x1x1y1y1, f);
+  // check denominators are correctly computed:
+  // den = 1 / (1 +- d * x1^2 * y1^2)
+  Provable.equal(
+    Field3,
+    one,
+    ForeignField.mul(x3Den, ForeignField.add(one, dx1x1y1y1, f), f)
+  ).assertTrue();
+  Provable.equal(
+    Field3,
+    one,
+    ForeignField.mul(y3Den, ForeignField.sub(one, dx1x1y1y1, f), f)
+  ).assertTrue();
+
+  ForeignField.assertMul(x3Num, x3Den, x3, f);
+  ForeignField.assertMul(y3Num, y3Den, y3, f);
```

Comment:
the method of first witnessing the denominators, then checking that they are computed correctly, and _then_ multiplying them with the numerators seems inefficient to me.

did you count the constraints created by these methods? and compare them with https://github.com/o1-labs/o1js/pull/1283?

my intuition says what I did in that PR is much more efficient, especially double but also add

Reply:
I get your point, seems reasonable. I am now purging the circuit, comparing against your PR and the Weierstrass gadget counterpart.

Reply:
make sure to have a constraint counting script! that will help you build intuition more quickly

like this one:
https://github.com/o1-labs/o1js/blob/a910d2c614c704b289c97ca5017d4b93fe49a577/src/examples/crypto/twisted-edwards.ts#L96-L103